### DISCUSSION TITLE
TPM fails to sign IKE_AUTH

### DISCUSSION THREAD

**Discussion by owensteam** [strongswan/strongswan#1177] [Q&A] [answered]
Hi there,

I get an error while trying to use the TPM for signing:

Aug 01 16:57:55 vpn1 charon-systemd[335425]: TPM 2.0 - Tss2_Sys_Sign failed: 0x0002d2
Aug 01 16:57:55 vpn1 charon-systemd[335425]: authentication of 'vpn1' (myself) with RSA signature failed

I followed the guide here:

https://docs.strongswan.org/docs/5.9/tpm/tpm2.html

so I have the key created, certreq created and signed, and then in my config I have a secrets section

secrets {
  token_ak_rsa {
    handle = 0x81010004
}
}

and the cert is here

    local {
      auth = pubkey
      certs = ak_rsa_cert_second.cer
      id = vpn1
    }

but when I try to connect I get the above error.

Any help massively appreciated

**Comment by strongX509** [comment]
Well, TPM2 error code  0x0002d2 means
````
 tpm:parameter(2):unsupported or incompatible scheme.
````
How did you create the AK signing key?

**Reply by owensteam** [reply]
thanks! Where did you get that error code from? I'm finding documentation difficult to come by

I just followed the guide in the strongswan doc:

```
sudo tpm2_createak -C 0x81010001 -G rsa -g sha256 -s rsassa -c ak_rsa_second.ctx -u ak_rsa_second.pub -n ak_rsa_second.name
sudo tpm2_evictcontrol -C o -c ak_rsa_second.ctx 0x81010004
sudo pki --req --type priv --keyid 0x81010004 --dn "CN=vpn1" --san "DNS:vpn1" --outform pem > ak_rsa_req_second.pem 
```

**Reply by strongX509** [reply]
Hmm, it seems that instead of the modern digital signature mode (`RSA_EMSA_PKCS1_SHA2_256`) a legacy **RSA signature** with a `SHA1` hash is negotiated.
````
Aug 01 16:57:55 vpn1 charon-systemd[335425]: authentication of 'vpn1' (myself) with RSA signature failed
````
Is your peer a Windows VPN client which does not support digital signature mode?

As a fix generate an AK with **-g sha1** and check if the signature succeeds. TPM 2.0 signature keys are always bound to a given hash algorithm. Thus an AK with a **-g sha256** profile can't be used to do a `SHA1` signature.

**Reply by owensteam** [reply]
wow thats fixed it, thank you so much.

Could I trouble you for a link to where I can find the error codes or how I find them using a manpage of some description?

for other logs (ike/ipsec proposals) I can see if they dont match and it tells me what to change, was there a way to get something similar for this do you know? Or did you just have to know the issue already?

**Reply by strongX509** [reply]
Use the `tpm2_rc_decode` command:
````
tpm2_rc_decode 0x0002d2
tpm:parameter(2):unsupported or incompatible scheme
````

**Comment by strongX509** [comment, marked as answer]
I think we can close this discussion
